ingestion/src/main/java/feast/source/common/ValueMapToFeatureRowTransform.java

```diff
@@ -0,0 +1,122 @@
+import feast.types.FeatureProto.Feature;
+import feast.types.FeatureRowProto.FeatureRow;
+import feast.types.ValueProto.Value;
+import feast.types.ValueProto.Value.ValCase;
+import java.util.Map;
+import java.util.Map.Entry;
+import org.apache.beam.sdk.extensions.protobuf.ProtoCoder;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.PCollection;
+
+public class ValueMapToFeatureRowTransform extends
+    PTransform<PCollection<Map<String, Value>>, PCollection<FeatureRow>> {
+
+  private String entity;
+  private Schema schema;
+
+  public ValueMapToFeatureRowTransform(String entity, Schema schema) {
+    this.entity = entity;
+    this.schema = schema;
+  }
+
+  @Override
+  public PCollection<FeatureRow> expand(PCollection<Map<String, Value>> input) {
+    return input.apply(ParDo.of(ValueMapToFeatureRowDoFn.of(entity, schema)))
+        .setCoder(ProtoCoder.of(FeatureRow.class));
+  }
+
+  public static class ValueMapToFeatureRowDoFn extends
+      DoFn<Map<String, Value>, FeatureRow> {
+
+    private String entityIdColumn;
+    private String timestampColumn;
+    private com.google.protobuf.Timestamp timestampValue;
+    private String entity;
+    private Map<String, Field> fields;
+
+    private ValueMapToFeatureRowDoFn(String entity, Map<String, Field> fields,
+        String entityIdColumn,
+        String timestampColumn,
+        com.google.protobuf.Timestamp timestampValue) {
+      this.entity = entity;
+      this.fields = fields;
+      this.entityIdColumn = entityIdColumn;
+      this.timestampColumn = timestampColumn;
+      this.timestampValue = timestampValue;
+    }
+
+    public static ValueMapToFeatureRowDoFn of(String entity, Schema schema) {
+      final Map<String, Field> fields = Maps.newHashMap();
+      for (Field field : schema.getFieldsList()) {
+        String displayName = !field.getName().isEmpty() ? field.getName() : field.getFeatureId();
+        fields.put(displayName, field);
+      }
+      return new ValueMapToFeatureRowDoFn(entity, fields, schema.getEntityIdColumn(),
+          schema.getTimestampColumn(), schema.getTimestampValue());
+    }
+
+    @ProcessElement
+    public void processElement(ProcessContext context) {
+      FeatureRow.Builder builder = FeatureRow.newBuilder();
+      try {
+        Map<String, Value> valueMap = context.element();
+        builder.setEntityName(entity);
+        for (Entry<String, Value> entry : valueMap.entrySet()) {
+          String name = entry.getKey();
+          Value value = entry.getValue();
+          Field field = fields.get(name);
+
+          if (value == null || value.getValCase().equals(ValCase.VAL_NOT_SET)) {
+            continue;
+          }
+
+          // A feature can only be one of these things
+          if (entityIdColumn.equals(name)) {
+            builder.setEntityKey(Values.asString(value).getStringVal());
```

Comment:
What happens if there are multiple columns with the same name?

Reply:
This is impossible if it's from a bq or json source. For csv, you'd have to specify multiple fields with the same name in the import spec, and that causes an IllegalArgumentException in the csv reader aborting the job.